rclc/src/functions.c

```diff
@@ -0,0 +1,347 @@
+
+    sub->user_callback(msg);
+  }
+
+  _rclc_spin_node_exit(&wait_set);
+}
+
+void
+rclc_spin_node(rclc_node_t * node)
+{
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t rc =
+    rcl_wait_set_init(&wait_set, node->subs_s, 0, 0, 0, 0, rcl_get_default_allocator());
+  if (rc != RCL_RET_OK) {
+    PRINT_RCL_ERROR(rclc_spin_node, rcl_wait_set_init);
+    return;
+  }
+
+  while (rcl_ok()) {
+    rc = rcl_wait_set_clear_subscriptions(&wait_set);
+    if (rc != RCL_RET_OK) {
+      PRINT_RCL_ERROR(rclc_spin_node, rcl_wait_set_clear_subscriptions);
+      _rclc_spin_node_exit(&wait_set);
+      return;
+    }
+
+    for (size_t i = 0; i < node->subs_s; ++i) {
+      rc = rcl_wait_set_add_subscription(&wait_set, &node->subs[i]->rcl_subscription);
+      if (rc != RCL_RET_OK) {
+        PRINT_RCL_ERROR(rclc_spin_node, rcl_wait_set_add_subscription);
+        _rclc_spin_node_exit(&wait_set);
+        return;
+      }
+    }
+
+    rc = rcl_wait(&wait_set, -1);
+    if (rc != RCL_RET_OK) {
+      PRINT_RCL_ERROR(rclc_spin_node, rcl_wait);
+      _rclc_spin_node_exit(&wait_set);
+      return;
+    }
+
+    for (size_t i = 0; i < wait_set.size_of_subscriptions; ++i) {
+      if (wait_set.subscriptions[i]) {
+        rclc_subscription_t * sub = NULL;
+
+        for (size_t j = 0; j < node->subs_s; ++j) {
+          if (&node->subs[j]->rcl_subscription == wait_set.subscriptions[i]) {
+            sub = node->subs[j];
+          }
+        }
+
+        if (sub == NULL) {
+          fprintf(stderr, "[rclc_spin_node] unable to find subscription in node.\n");
+          _rclc_spin_node_exit(&wait_set);
+        }
+
+        void * msg = ZERO_ALLOCATE(sub->type_support.size_of);
+
+        rc = rcl_take(wait_set.subscriptions[i], msg, NULL);
+        if (rc != RCL_RET_OK) {
+          PRINT_RCL_ERROR(rclc_spin_node, rcl_take);
+          _rclc_spin_node_exit(&wait_set);
+          return;
+        }
+
+        sub->user_callback(msg);
+      }
+    }
+  }
+
+  _rclc_spin_node_exit(&wait_set);
+}
+
+rclc_node_t *
+rclc_create_node(const char * name, const char * namespace_)
+{
+  rclc_node_t * rclc_node = ALLOCATE(sizeof(rclc_node_t));
+  rclc_node->rcl_node = rcl_get_zero_initialized_node();
+  rclc_node->subs = NULL;
+  rclc_node->subs_s = 0;
+
+  rcl_node_options_t node_ops = rcl_node_get_default_options();
+  rcl_ret_t rc = rcl_node_init(&rclc_node->rcl_node, name, namespace_, &node_ops);
+  if (rc != RCL_RET_OK) {
+    PRINT_RCL_ERROR(rclc_create_node, rcl_node_init);
+    return NULL;
+  }
+
+  return rclc_node;
+}
+
+rclc_ret_t
+rclc_destroy_node(rclc_node_t * node)
+{
+  rcl_ret_t rc = rcl_node_fini(&node->rcl_node);
+  if (rc != RCL_RET_OK) {
+    PRINT_RCL_ERROR(rclc_destroy_node, rcl_node_fini);
+  }
+  DEALLOCATE(node);
+  return rc;
+}
+
+rclc_publisher_t *
+rclc_create_publisher(
+  rclc_node_t * node,
+  const rclc_type_support_t type_support,
+  const char * topic_name,
+  size_t queue_size)
+{
+  (void)queue_size;
```

Comment:
Can you please add a TODO here so that we remember to take this parameter into account in a future iteration